<!DOCTYPE HTML>
<!--
	Ion by TEMPLATED
	templated.co @templatedco
	Released for free under the Creative Commons Attribution 3.0 license (templated.co/license)
-->
<html xmlns="http://www.w3.org/1999/html">
<head>
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-140303493-1"></script>
    <script>
        window.dataLayer = window.dataLayer || [];
        function gtag(){dataLayer.push(arguments);}
        gtag('js', new Date());

        gtag('config', 'UA-140303493-1');
    </script>

    <title>Flint - Getting Started</title>
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <meta name="description" content="" />
    <meta name="keywords" content="" />
    <!--[if lte IE 8]><script src="js/html5shiv.js"></script><![endif]-->
    <script src="js/jquery.min.js"></script>
    <script src="js/skel.min.js"></script>
    <script src="js/skel-layers.min.js"></script>
    <script src="js/init.js"></script>

    <!-- Google Code Prettify -->
    <script src="https://cdn.jsdelivr.net/gh/google/code-prettify@master/loader/run_prettify.js"></script>

    <noscript>
        <link rel="stylesheet" href="css/skel.css" />
        <link rel="stylesheet" href="css/style.css" />
        <link rel="stylesheet" href="css/style-xlarge.css" />
    </noscript>

</head>
<body id="top">

<!-- Header -->
<header id="header" class="skel-layers-fixed">
    <h1><a href="index.html">Flint</a></h1>
    <nav id="nav">
        <ul>
            <li><a href="index.html" title="Home Page">Home</a></li>
            <li><a href="getting_started.html" title="Getting Started">Getting Started</a></li>
            <li><a href="documentation.html" title="Documentation">Documentation</a></li>
            <li><a href="faq.html" title="FAQ">FAQ</a></li>
            <li><a href="https://github.com/camilo-v/flint" target="_blank" title="GitHub">GitHub</a></li>
        </ul>
    </nav>
</header>

<!-- Main -->
<section id="main" class="wrapper style1">
    <header class="major">
        <h2>Getting Started</h2>
        <p>Analyzing Large Metagenomics Datasets</p>
    </header>
    <div class="container">
        <div class="row">
            <div class="8u">
                <section>
                    <p>Flint is a metagenomics profiling pipeline that is built on top of the <a href="https://spark.apache.org" target="_blank">Apache Spark</a> framework, and is designed for fast profiling of metagenomic samples against a large collection of reference genomes. Flint takes advantage of Spark's built-in parallelism and streaming engine architecture to quickly map reads against a large reference collection of bacterial genomes.</p>

                    <h3>Required üö©</h3>
                    <p>The Flint software is designed to run on <a href="https://aws.amazon.com" title="Amazon Web Services" target="_blank">AWS</a>, and requires a couple of assets to be <em>in place</em> before you can run it. Following is a list of the tools &amp; items that you will need in order to setup and run a Flint cluster:</p>

                    <ol>
                        <li>An <a href="https://aws.amazon.com" title="Amazon Web Services" target="_blank">AWS Account</a>.</li>
                        <li>A copy of the partitioned <a href="#bacterial_genomes">genomes collection</a>.</li>
                        <li>A copy of the Flint <a href="https://github.com/camilo-v/flint" title="Flint Project" target="_blank">source code</a>.</li>
                        <li>A terminal emulator for your computer (<a href="https://en.wikipedia.org/wiki/Terminal_(macOS)" title="macOS Terminal" target="_blank">Terminal.app</a>, <a href="https://www.iterm2.com" title="iTerm2" target="_blank">iTerm2</a>, <a href="https://www.chiark.greenend.org.uk/~sgtatham/putty/" title="PuTTY" target="_blank">PuTTY</a>, etc.)</li>
                        <li>An SFTP client (<a href="https://panic.com/transmit/" title="Transmit" target="_blank">Panic's Transmit</a>, <a href="https://filezilla-project.org" title="FileZilla" target="_blank">FileZilla</a>, etc.).</li>
                    </ol>

                </section>

                <hr class="major" />

                <section>
                    <h2>Deployment, Configuration, and Installation</h2>
                    <p>Flint runs on <a href="https://spark.apache.org" target="_blank">Spark</a>, and the current version is tuned for <a href="https://aws.amazon.com/emr/" target="_blank">Amazon's EMR service</a>. While the Amazon-specific code is minimal, we are not yet supporting Spark clusters outside of EMR. A Flint project consists of multiple pieces (outlined below), and you will need to stage the genome reference assets and machine configurations before launching a cluster. The pieces are:</p>
                    <ol>
                        <li><strong><a href="#assets">‚úÖ Asset Staging: </a></strong> Upload the bacterial reference genomes to an accessible data bucket in Amazon's S3 storage, along with two configuration scripts for the cluster.</li>
                        <li><strong><a href="#emr">üéõ EMR Cluster Launch: </a></strong> Launch an EMR cluster that will be configured with bootstrap actions and steps that you staged in the previous step.</li>
                        <li><strong><a href="#access">üì° Accessing the Cluster:</a></strong> Connect to the cluster that you just created. You can connect through SSH or through a SFTP client.</li>
                        <li><strong><a href="#source">‚öôÔ∏è Source Code Deployment: </a></strong> Upload the main Flint python script, along with utilities for copying the reference shards into each cluster node, and a template of the <samp>spark-submit</samp> resource file, into the cluster's master node.</li>
                        <li><strong><a href="#terminate">üõë Terminating a Cluster:</a></strong></strong> Any cloud provider will charge you for the time you use, so its critical that you terminate the clusters after you are done using them.</li>
                    </ol>
                    <p>The following section(s) contain details on each of the above. If you have any questions, comments, and/or queries, please <a href="https://github.com/camilo-v/flint" title="Flint Project" target="_blank">get in touch</a>.</p>
                    <hr class="major" />
                </section>


                <!-- ----------------------------------------- Asset Staging -------------------------------------- -->
                <section>
                    <h2><a name="assets" class="anchor"></a>‚úÖ Asset Staging</h2>
                    <p>This section involves two (2) parts, briefly described below and in detail after:
                        <ol>
                            <li>Upload of the partitioned bacterial genomes collection into a location (in AWS) that is accessible by the worker machines in the EMR cluster.</li>
                            <li>
                                Upload of two configuration scripts that the EMR launch procedure will use to provision the cluster.
                            </li>
                        </ol>
                    </p>

                    <!-- -------------------------- Partitioned Bacterial Genomes ------------------------------ -->
                    <br>
                    <h3><a name="bacterial_genomes" class="anchor"></a>1. Partitioned Bacterial Genomes</h3>
                    <p>
                        The bacterial genomes are the largest asset that you will deposit in your S3 bucket. In the Flint publication, we used a bacterial genome collection from <a href="https://bacteria.ensembl.org/index.html" title="Ensembl Bacterial" target="_blank">Ensembl (v.41)</a> that was partitioned into 64 shards, with each shard going to a single machine in the cluster. Below we provide genome collections for <a href="https://bacteria.ensembl.org/index.html" title="Ensembl Bacterial" target="_blank">Ensembl (v.41)</a> partitioned into 128, 256, and 512 shards, which can be used in <a href="https://spark.apache.org" target="_blank">Spark</a> clusters of those sizes.
                    </p>

                    <p>
                        ‚ö†Ô∏è <strong>Note:</strong> <a href="https://aws.amazon.com/s3/pricing/">S3 storage prices</a> change, so please check that the bucket you are uploading to has the capacity to host the genome collection.
                    </p>

                    <p>
                        Each partitioned collection shard includes subdirectories that contain the corresponding <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#small-and-large-indexes" target="_blank" title="Bowtie2 Index">Bowtie2 index files (*.bt2)</a>. You will need to copy all the shards for a given cluster size (64, 128, etc.) into an S3 bucket that is accessible to your EMR cluster, i.e., to an S3 bucket in your account.
                    </p>

                    <p>
                        ‚ùóÔ∏èThe Flint source code <a href="documentation.html" title="Documentation">includes a script</a> that will copy all the shards into each worker node from your S3 bucket. This only needs to be done once per cluster setup.
                    </p>

                    <!-- ------------------------------------- Downloading ------------------------------------- -->
                    <br>
                    <h4>Downloading within AWS ‚¨áÔ∏è</h4>
                    <p>
                        You will need to use the <a href="https://aws.amazon.com/cli/" target="_blank" title="AWS CLI">AWS CLI</a> utilitiy to sync the genomes from our buckets into a bucket in your account. The collections are available below:
                    </p>

                    <ol>
                        <li><a href="https://s3.amazonaws.com/flint-bacterial-genomes/ensembl/41/indices/partitions_64" title="64" target="_blank">64</a>: <samp>s3://flint-bacterial-genomes/ensembl/41/indices/partitions_64</samp></li>
                        <li><a href="https://s3.amazonaws.com/flint-bacterial-genomes/ensembl/41/indices/partitions_128" title="128" target="_blank">128</a>: <samp>s3://flint-bacterial-genomes/ensembl/41/indices/partitions_128</samp></li>
                        <li><a href="https://s3.amazonaws.com/flint-bacterial-genomes/ensembl/41/indices/partitions_256" title="256" target="_blank">256</a>: <samp>s3://flint-bacterial-genomes/ensembl/41/indices/partitions_256</samp></li>
                        <li><a href="https://s3.amazonaws.com/flint-bacterial-genomes/ensembl/41/indices/partitions_512" title="512" target="_blank">512</a>: <samp>s3://flint-bacterial-genomes/ensembl/41/indices/partitions_512</samp></li>
                    </ol>

                    <p>
                        An example <samp>sync</samp> command to copy the genomes using <a href="https://aws.amazon.com/cli/" target="_blank" title="AWS CLI">AWS CLI</a> is below, you just have to substitute <samp>my-bucket</samp> with the path of your S3 bucket:
                    </p>

                    <code class="prettyprint lang-bsh">
aws s3 sync s3://flint-bacterial-genomes/ensembl/41/indices/partitions_64 s3://my-bucket/partitions_64 --request-payer requester
                    </code>

                    <p>
                        ‚ö†Ô∏è As always with AWS, there is a cost for transferring data. Your AWS account will be charged for the cost of transferring the genomes into your bucket, as specified by <samp>--request-payer requester</samp>.
                    </p>

                    <p>
                        ‚ö†Ô∏è Do note that the syncing process can take some time, so make sure that when you issue the <samp>sync</samp> command, the terminal window in which you do so remains open, or you <a href="https://serverfault.com/questions/41959/how-to-send-jobs-to-background-without-stopping-them" title="Background Jobs" target="_blank">send the job to the background</a>.
                    </p>

                    <br>

                    <h4>Downloading with SFTP ‚¨áÔ∏è</h4>
                    <p>
                        Due to the size of the partitioned collections, we are not supporting SFTP transfer at this time.
                    </p>


                    <br><br>

                    <h3><a name="cluster_configuration" class="anchor"></a>2. Cluster Configurations</h3>
                    <p>
                        The cluster is configured with two (2) scripts that are provided to the EMR service when a cluster is created. These two files need to be uploaded to an S3 bucket that is accessible to EMR so that you can select them in the cluster creation step.
                    </p>

                    <p> The first script is a JSON file that is used in the first step of the cluster creation, "<a href="#software_steps" title="Software and Steps">Software and Steps</a>", below &mdash; this is the JSON file that you specify in the "Software Settings" part. The second is a BASH file that defines what software will be loaded into each machine of the cluster (master and worker nodes). The BASH script defines environment variables, sets up the required directories, and loads the <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">Bowtie2</a> aligner. This BASH file is used the "<a href="#general_cluster_settings" title="General Cluster Settings">General Cluster Settings</a>" steps below, and is the file that is pointed to by the "Bootstrap Actions" in the configuration.
                    </p>


                    <!-- ------------------------------------- EMR Software Settings --------------------------- -->
                    <br>
                    <h4><a name="emr_json_software_settings" class="anchor"></a>EMR JSON Software Settings</h4>
                    <p>The EMR software settings are simple, and they specify the particulars that will be used to spin-up a Spark cluster. Note that the below settings are specific to the EMR <a href="https://aws.amazon.com/ec2/instance-types/" title="Amazon EC2 Instance Types" target="_blank">machine instances</a> of type <samp>c4.2xlarge</samp>, these instances contain 8 <a href="https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-optimize-cpu.html" title="Optimizing CPU Options" target="_blank">vCPU</a> and 15 GB of memory.</p>
                    <script src="https://gist.github.com/camilo-v/710dd9d93eea2d77a76243ce35a0ed35.js"></script>

                    <br>
                    <h4><a name="emr_bootstrap_action" class="anchor"></a>EMR Bootstrap Action</h4>
                    <p>The bootstrap action is a little more complicated, and it defines all the software that each machine will contain. The BASH script for this is located <a href="amazon-emr/bootstrap_actions/app-setup.sh" title="app-setup.sh">here</a>, or in the "amazon-emr" directory of the Flint project, under the folder "bootstrap actions".</p>

                    <p>The <a href="amazon-emr/bootstrap_actions/app-setup.sh" title="app-setup.sh">script</a> is also available as a <a href="https://gist.github.com/camilo-v/a36d778bcfdb500be04e08bf51efe072" title="app-setup.sh gist">github gist.</a></p>

                    <p>You will need to make a copy of this script, change line number "49" so that it points to the root bucket name in which you uploaded the file so that the EMR setup can pick it up. The variable <samp>SOURCE_BUCKET_NAME=s3-bucket-name</samp> in <strong>line 49</strong> defines the bucket name in which you are staging your assets in.</p>

                    <code class="prettyprint lang-bsh">
# Edit line 49 to match your bucket's name.
SOURCE_BUCKET_NAME="s3-bucket-name"
                    </code>



                </section>

                <hr class="major" />

                <br>

                <!-- --------------------------------- Cluster Launch ---------------------------------- -->
                <section>
                    <h2><a name="emr" class="anchor"></a>üéõ EMR Cluster Launch</h2>
                    <p>
                        Flint is currently optimized for running on AWS, and on the EMR service specifically. You will need an AWS account to launch an EMR cluster, so familiarity with AWS and EMR is ideal. If you are not familiar with AWS, you can start with <a href="https://aws.amazon.com/getting-started/" target="_blank"> this tutorial</a>. Please note that the following steps, and screenshots, all assume that you have successfully signed into your <a href="https://aws.amazon.com/console/" target="_blank">AWS Management Console</a>.
                    </p>

                    <a href="#" class="image fit"><img src="images/getting_started/aws-console.png" alt="aws management console" title="aws management console"/></a>

                    <p>
                        In the management console main page, you can search for "EMR"; this will take you to the EMR dashboard, which will load with the clusters you have previously created (if any). Click on the <strong>Create Cluster</strong> button to start.
                    </p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-1.png" alt="create cluster" title="create cluster"/></a>

                    <p>
                        This will take you to the <strong>"Create Cluster - Quick Options"</strong> configuration page. You can launch a cluster in this page right away, but we'll go ahead and click on the <strong>"Go to advanced options"</strong> link to launch a cluster with some specific changes required for Flint.
                    </p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-2.png" alt="advanced options" title="advanced options"/></a>

                    <p>
                        The "<strong>Advanced Options</strong>" has four (4) steps that you have to configure, these are:
                        <ol>
                            <li><a href="#software_steps" title="Software and Steps">Software and Steps</a></li>
                            <li><a href="#hardware" title="Hardware">Hardware</a></li>
                            <li><a href="#general_cluster_settings" title="General Cluster Settings">General Cluster Settings</a></li>
                            <li><a href="#security" title="Security">Security</a></li>
                        </ol>
                        We'll go through each one, and configure them step-by-step below.
                    </p>

                    <!-- ------------------------------------ Software and Steps ------------------------------------ -->
                    <h3><a name="software_steps" class="anchor"></a> 1. Software and Steps</h3>
                    <p>In this step of the configuration, we are going to tell the EMR service the basic software that our cluster will run on. Flint was tested and developed on EMR release "<strong>emr-5.20.0</strong>" but you should be able to select the latest release. If you see that you get an error with a particular release configuration, <a href="https://github.com/camilo-v/flint/issues" target="_blank">please submit an issue</a>.</p>
                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options_1.png" alt="Software and Steps" title="Software and Steps"/></a>
                    <p>
                        Make sure the highlighted items in the above screenshot are selected, and then copy the following code for the software settings configuration:
                        <script src="https://gist.github.com/camilo-v/710dd9d93eea2d77a76243ce35a0ed35.js"></script>
                        You can also view the <a href="https://gist.github.com/camilo-v/710dd9d93eea2d77a76243ce35a0ed35" target="_blank">gist in github</a>. These settings are the ones described in the <a href="#emr_json_software_settings">EMR JSON</a> settings in the "<a href="#cluster_configuration" title="Cluster Configuration">Cluster Configuration</a>" above.
                        Click the "<strong>Next</strong>" button to proceed to the next step.
                    </p>

                    <!-- ------------------------------------------- Hardware ------------------------------------ -->
                    <h3><a name="hardware" class="anchor"></a>2. Hardware</h3>
                    <p>In this step we'll configure the <a href="https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-supported-instance-types.html" target="_blank">type and number of machines</a> that the cluster will contain. Be very careful when selecting the instance types, as each type has a corresponding cost, and if you are not careful you can be charged a lot. You can view EMR instance type <a href="https://aws.amazon.com/emr/pricing/" target="_blank">costs</a> by going to <a href="https://aws.amazon.com/emr/pricing/" target="_blank">this page</a>.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options_2_1.png" alt="Hardware" title="Hardware"/></a>

                    <p>In the Flint paper, we used instances of type <samp>c4.2xlarge</samp> for both the "Master" and "Worker" nodes. You can change the instance types by clicking on the "<strong>pencil</strong> icon (circled <a style="color: red">red</a>) in the above screenshot, and then changing the instance type. The worker nodes are configured in the "<strong>Core</strong>" section, and you can change the number of worker nodes that you need &mdash; here, we selected 64. Note that this has to match the <strong>number of partitions of your genome references</strong>.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options_2_2.png" alt="Hardware - Costs" title="Hardware - Costs"/></a>

                    <p>After you have selected the instance type, and number of them, you can obtain a lower price per machine by selecting the "<strong>Spot</strong>" purchasing option (see above). Hovering over the "<strong>information</strong>" icon should bring up a menu with the list of current spot-market prices for the instance you selected. In the paper, the <samp>c4.2xlarge</samp> instances were priced at $0.123 USD per hour per machine. In this page you can also select the maximum amount that you are willing to pay for an instance, in our case we specified $0.159 USD.</p>

                    <!-- ------------------------------------- General Cluster Settings ------------------------------------ -->
                    <h3><a name="general_cluster_settings" class="anchor"></a>3. General Cluster Settings</h3>
                    <p>In this section we will configure the actual software that each machine will contain. Custom Python libraries, the <a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml" target="_blank">Bowtie2</a> aligner, etc. are all loaded in this step. To start you can give your cluster a name, and select a location were deployment and runtime logs will be stored -- note these locations have to be in a S3 bucket that you have acess to.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options_3_1.png" alt="General Cluster Settings" title="General Cluster Settings"/></a>

                    <h4>Bootstrap Actions</h4>
                    <p>The critical section in this step is the Bootstrap Actions &mdash; this is the part in which you provide a script that will provision each machine with the required software. In the "<strong>Bootstrap Actions</strong> section, select the "<strong>Custom action"</strong> option in the "<strong>Add bootstrap action</strong>" dropdown. This will enable the "<strong>Configure and add</strong>" button, which will allow you to select a location in an S3 bucket that contains the script with the steps to provision each machine.
                    </p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options_3_2.png" alt="General Cluster Settings - Bootstrap Actions" title="General Cluster Settings - Bootstrap Actions"/></a>

                    <p>The script that provisions each machine is a BASH shell-script, and you have to have uploaded it to an accessible S3 location that you can point to. This BASH script is the file from the "<a href="#cluster_configuration" title="Cluster Configuration">Cluster Configuration</a>" section above, and the one under the "<a href="#emr_bootstrap_action" title="Bootstrap Action">Bootstrap Action</a>" part.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options_3_3.png" alt="General Cluster Settings - Bootstrap Action Script" title="General Cluster Settings - Bootstrap Action Script"/></a>

                    <p>Give the action a name, and then click on the "Add" button. Once the action has been successfully added, click "Next" to proceed to the next section.</p>


                    <!-- ------------------------------------- Security ------------------------------------ -->
                    <h3><a name="security" class="anchor"></a>4. Security</h3>
                    <p>The last step involves setting the access credentials to the cluster. In the "<strong>Security Options</strong>" section you have to select the <strong>EC2 key pair</strong> for the <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html">IAM User</a> that will be accessing the cluster. Double-check that the Master and Core (worker) nodes have the security group that includes the IAM user, and you should be good to go. Click on the <strong>Create Cluster</strong> button to finish.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options_4.png" alt="Security" title="Security"/></a>

                    <!-- ------------------------------------- Finish ------------------------------------ -->
                    <h3>Finish</h3>
                    <p>After you click the <strong>Create Cluster</strong> button you will be taken to the details page of the cluster you just created. The page will display status messages and give you an idea of the progress of the launch.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options-5.png" alt="Finish" title="Finish"/></a>

                    <p>Depending on the resources (instance types) that you requested, along with spot-market availability (if you used it), the cluster should be ready in about 8-15 minutes. The cluster is ready to use when the status message: "<span style="color: green">Waiting</span>" Cluster ready after last step completed." appears, as well as both the Master and Core nodes show a status of "<span style="color: green">Running</span>".</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-3-options-6.png" alt="Finish - connections" title="Finish  - connections"/></a>

                    <p>Make a note of the entry for "<strong>Master public DNS</strong>" as you will need it to connect to the cluster later on.</p>

                </section>

                <hr class="major" />

                <br>


                <!-- ---------------------------------- Accessing the Cluster ------------------------------------- -->

                <section>
                    <h2><a name="access" class="anchor"></a>üì° Accessing the Cluster</h2>
                    <p>Once you have successfully launched the cluster you can connect to it via <strong>SSH</strong> or <strong>SFTP</strong> to launch jobs and upload files respectively. The sections below outline how to connect to the Master node, and also how to access diagnostic and monitoring dashboards.</p>

                    <br>

                    <h3>SSH</h3>
                    <p>The URL for connecting to the cluster is displayed next to the "<strong>Master public DNS</strong>" label, or you can click on the "<strong>SSH</strong>" link to bring up a dialog with the exact command you can copy/paste into your SSH client (terminal.ap, etc.).</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-4-1.png" alt="Connecting" title="Connecting"/></a>

                    <p><strong>Note:</strong>You will need a copy of the same IAM User certificate you created the cluster with to connect to the cluster with SSH. In the command that EMR gives you, the <samp>-i</samp> option to the <samp>ssh</samp> command specifies the path to the certificate.</p>

                    <p>Just copy/paste the url that EMR gives you into an SSH command in an SSH client such as Terminal.app in macOS, specify the certificate with the <samp>-i</samp> parameter and make sure that you have the user as <samp>hadoop</samp> &mdash; its the default EMR username.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-4-2.png" alt="Connecting - SSH" title="Connecting - SSH"/></a>

                    <p>Since this is the first time we are connecting to the cluster, SSH will ask us if we want to add the cluster's IP address to the list of known hosts, go ahead and type "<strong>yes</strong>" and you will connect.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-4-3.png" alt="Connecting - SSH done" title="Connecting - SSH done"/></a>

                    <p>Once you see the ASCII logo for "<strong>EMR</strong>" you are successfully connected to the Master node. From here, you can jump to any of the other Worker machines by using the private key certificate that you launched the cluster with and used to connect to the master node.</p>

                    <br>

                    <h3>SFTP</h3>
                    <p>You can upload files to the cluster through an SFTP client such as <a href="https://filezilla-project.org" target="_blank">Filezilla</a> or <a href="https://panic.com/transmit/" target="_blank">Transmit 5</a>. If you are using Transmit on macOS, the address of the cluster should be the same that you used in the SSH command, or the one that EMR gives you next to the "<strong>Master public DNS</strong>" label. The <strong>User Name</strong> should be "hadoop" and for the password just use the same private key certificate (IAM user) that you setup the cluster with.</p>

                    <a href="#"><img src="images/getting_started/emr-4-4.png" height="auto" width="400px" alt="Connecting - SFTP" title="Connecting - SFTP"/></a>

                    <p>After you connect, you will be taken to the home directory of the hadoop user.</p>

                    <br>

                    <h3>Ganglia Instrumentation</h3>
                    <p>You can monitor the health of your cluster, as well as track resource usage with <a href="http://ganglia.sourceforge.net" target="_blank">Ganglia</a>. The Ganglia dashboard is not readily accessible, and you need to do some setup before you can connect to it. The link in the titled "<strong>Enable Web Connection</strong> has detailed instructions on how to do this, but briefly: you have to establish an SSH tunnel from an SSH client (Terminal.app in macOS), and setup <a href="https://addons.mozilla.org/en-US/firefox/addon/foxyproxy-standard/" target="_blank">FoxyProxy</a> in a supported Web Browser. </p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-4-5.png" alt="Connecting - Web Connection" title="Connecting - Web Connection"/></a>

                    <p>Establishing an SSH tunnel, and FoxyProxy instructions:</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-4-6.png" alt="Connecting - Web Connection details" title="Connecting - Web Connection details"/></a>

                    <p>Once you have setup the SSH tunnel and FoxyProxy, then you can access the Ganglia dashboard by adding <samp>/ganglia</samp> to the end of the "<strong>Master public DNS</strong>" address. Note that this has to be done in the same browser that you installed and configured FoxyProxy, and after you started the SSH tunnel. Order matters, so first you start the SSH tunnel, and then you access the web interface.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-4-7.png" alt="Connecting - Ganglia" title="Connecting - Ganglia"/></a>

                    <p>Ganglia information for all nodes.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-4-8.png" alt="Connecting - Ganglia graph" title="Connecting - Ganglia graph"/></a>

                </section>

                <hr class="major" />

                <br>

                <!-- ---------------------------------- Source Code Deployment ------------------------------------ -->
                <section>
                    <h2><a name="source" class="anchor"></a>‚öôÔ∏è Source Code Deployment</h2>
                    <p>
                        At this point you have launched a cluster in EMR, and have a partitioned genome collection in an S3 bucket that is accessible to the cluster. The next step is to upload the Flint source code to the cluster so that you can start running jobs. If you have not done so already, go ahead an grab a copy of the <a href="https://github.com/camilo-v/flint" title="Source Code" target="_blank">Flint source code in GitHub</a>.
                    </p>
                    <p>
                        The code, all of it, will need to be uploaded to the root directory of the <samp>hadoop</samp> user in the cluster. You can review the section <a href="#access" title="Accessing the Cluster">Accessing the Cluster</a> for instructions on how to connect to your cluster.
                    </p>
                    <p>
                        Once the source code has been uploaded, you are ready to start running jobs. The <a href="documentation.html" title="Documentation">documentation page</a> has example code, details on how to run Flint job, as well as API references of the relevant functions. You can also review the basics on <a href="https://spark.apache.org/docs/latest/submitting-applications.html" title="Spark, Submitting Applications" target="_blank">Submitting Applications</a> in Spark.
                    </p>

                    <p>
                        <a href="documentation.html" title="Documentation">Read Flint Documentation &rightarrow;</a>
                    </p>

                </section>

                <hr class="major" />

                <br>

                <!-- ---------------------------------------- Terminating ---------------------------------------- -->
                <section>
                    <h2><a name="terminate" class="anchor"></a>üõë Terminating a Cluster</h2>
                    <p>In AWS, Amazon <a href="https://aws.amazon.com/emr/pricing/" target="_blank">charges you for the time</a> you use the cluster, and if you do not terminate the cluster after you are done with it, the bill can be quite large. Terminating a cluster will stop all jobs and any code and/or resources you uploaded to the cluster will be lost. Terminating the cluster is easy, in the cluster details page, just click on the "<strong>Terminate</strong>" button.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-6-1.png" alt="Terminate" title="Terminate"/></a>

                    <p>Clicking on the "<strong>Terminate</strong>" button will bring up the "Terminate" dialog. Click on "<strong>Terminate</strong>" to stop the cluster.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-6-2.png" alt="Terminate - Dialog" title="Terminate - Dialog"/></a>

                    <p>When the status of the cluster is displayed as <strong>Terminated</strong>, the cluster is no longer available, but note that it will still be visible in the cluster list in the main EMR dashboard.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-6-3.png" alt="Terminate - Terminated" title="Terminate - Terminated"/></a>

                    <p>A cool part of having the cluster details available after you terminated a cluster is that you can "<strong>Clone</strong>" it (relaunch a new cluster with the same parameters and configurations of a previous cluster) or copy a AWS CLI command to relaunch a cluster right from the command line.</p>

                    <a href="#" class="image fit"><img src="images/getting_started/emr-6-4.png" alt="Terminate - Relaunch Options" title="Terminate - Relaunch Options"/></a>

                </section>

                <hr class="major" />

                <p><em>Code forth and prosper.</em> üññ</p>
                <br>

            </div>
            <div class="4u">
                <section>
                    <h3>Flint Metagenomics</h3>
                    <p>Flint is a metagenomics profiling pipeline that is built on top of the <a href="https://spark.apache.org" target="_blank">Apache Spark</a> framework, and is designed for fast real-time profiling of metagenomic samples against a large collection of reference genomes.</p>
                    <ul class="actions">
                        <li><a href="getting_started.html" class="button alt">Learn More</a></li>
                    </ul>
                </section>
                <hr />
                <section>
                    <h3>Resources üóÑ</h3>
                    <ul class="alt">
                        <li><a href="getting_started.html">Getting Started with Flint</a></li>
                        <li><a href="documentation.html" title="Documentation">Documentation</a></li>
                        <li><a href="faq.html">Frequently Asked Questions</a></li>
                    </ul>
                </section>
                <section>
                    <h3>Documentation üìïüìó</h3>
                    <ul class="alt">
                        <li><a href="#">Overview</a></li>
                        <li><a href="#">Sample Code</a></li>
                        <li><a href="#">Partitioned Genomes</a></li>
                        <li><a href="#">Learn More...</a></li>
                    </ul>
                </section>

                <section>
                    <h3>Publication üè∑</h3>
                    <p>Flint can be referenced by using the citation:</p>
                    <ul class="alt">
                        <li><strong>Valdes</strong>, <strong>Stebliankin</strong>, <strong>Narasimhan</strong> (2019), <a href="#">Large Scale Microbiome Profiling in the Cloud</a>, <em>ISMB 2019, in review</em>.</li>
                    </ul>
                </section>
            </div>
        </div>
    </div>
</section>

<!-- Footer -->
<footer id="footer">
    <div class="container">
        <div class="row double">
            <div class="6u">
                <h2>Flint - Metagenomics</h2>
                <p>Flint is a metagenomics profiling pipeline built on top of <a href="https://spark.apache.org" target="_blank">Apache Spark</a>, and is designed for fast real-time processing of samples against large collections of reference genomes.</p>
                <ul class="icons">
                    <li><a href="#" class="icon fa-twitter"><span class="label">Twitter</span></a></li>
                    <li><a href="#" class="icon fa-facebook"><span class="label">Facebook</span></a></li>
                    <li><a href="#" class="icon fa-instagram"><span class="label">Instagram</span></a></li>
                    <li><a href="#" class="icon fa-linkedin"><span class="label">LinkedIn</span></a></li>
                </ul>
            </div>
            <div class="6u">
                <div class="row collapse-at-2">
                    <div class="6u">
                        <h3>Flint</h3>
                        <ul class="alt">
                            <li><a href="https://github.com/camilo-v/flint" title="GitHub Repository">GitHub</a></li>
                            <li><a href="documentation.html" title="Documentation">Documentation</a></li>
                            <li><a href="getting_started.html" title="Getting Started">Getting Started</a></li>
                            <li><a href="faq.html" title="Frequently Asked Questions">FAQ</a></li>
                        </ul>
                    </div>
                    <div class="6u">
                        <h3>&nbsp;</h3>
                        <ul class="alt">
                            <li><a href="#">Publication</a></li>
                            <li><a href="http://biorg.cs.fiu.edu" title="BioRG lab at FIU">BioRG Lab</a></li>
                            <li><a href="https://www.cis.fiu.edu" title="FIU">FIU Computer Science</a></li>
                            <li><a href="https://github.com/camilo-v/flint/issues" title="Bugs">Bug Report</a></li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
        <ul class="copyright">
            <li>&copy; Camilo Valdes. All rights reserved.</li>
            <li>Design: <a href="http://templated.co">TEMPLATED</a></li>
            <li>Images: <a href="http://unsplash.com">Unsplash</a></li>
        </ul>
    </div>
</footer>

</body>
</html>
